### PR TITLE
Add lock while creating engine with non-CONNECTION mode

### PR DESCRIPTION
**Pull request by Honglun**
### _Why are the changes needed?_

We may create multiple SparkContext in non-CONNECTION ShareLevel.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/HighAvailabilityConf.scala [hunk 1]
```diff
@@ -99,4 +99,11 @@ object HighAvailabilityConf {
     .timeConf
     .checkValue(_ > 0, "Must be positive")
     .createWithDefault(Duration.ofSeconds(120).toMillis)
+
+  val HA_ZK_ENGINE_LOCK_TIMEOUT: ConfigEntry[Long] =
+    buildConf("ha.zookeeper.engine.lock.timeout")
+      .doc("The timeout(s) of a lock when a engine created")
+      .version("1.0.0")
+      .timeConf
+      .createWithDefault(Duration.ofSeconds(30).toMillis)
```
Comment (ulysses-you) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/HighAvailabilityConf.scala:108]: The default timeout number I think should be same with `session.engine.initialize.timeout` which is the max timeout before fail.

**Review by ulysses-you — COMMENTED** [review]
@Honglun thank you for working on this. I left some comments.

It is a real issue if the same user open a new session more than once at the same time using `USER` mode. cc @yaooqinn

**Comment by Honglun** [comment]
@ulysses-you Thanks, I will update the code latter.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala [hunk 2]
```diff
@@ -342,6 +343,31 @@ object ServiceDiscovery extends Logging {
     }
     serviceNode
   }
+
+  def withLock(
+      zkClient: CuratorFramework,
+      lockPath: String,
+      lockTimeOutSeconds: Long)(f: => Unit): Unit = {
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:350]: it's milliseconds

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala [hunk 3]
```diff
@@ -342,6 +343,31 @@ object ServiceDiscovery extends Logging {
...
+    var lock: InterProcessSemaphoreMutex = null
+    try {
+      try {
+        lock = new InterProcessSemaphoreMutex(zkClient, ZKPaths.makePath(lockPath, "lock"))
+        lock.acquire(lockTimeOutSeconds, TimeUnit.SECONDS)
+      } catch {
+        case e: Exception => throw new KyuubiException(s"Lock failed on path [$lockPath]", e)
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:357]: KyuubiSQLException

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala [hunk 4]
```diff
@@ -342,6 +343,31 @@ object ServiceDiscovery extends Logging {
     }
     serviceNode
   }
+
+  def withLock(
+      zkClient: CuratorFramework,
+      lockPath: String,
+      lockTimeOutSeconds: Long)(f: => Unit): Unit = {
+    var lock: InterProcessSemaphoreMutex = null
+    try {
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:352]: Can we simply use `scala.util.Using`? @pan3793

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/SQLEngineAppName.scala [hunk 5]
```diff
@@ -32,6 +32,12 @@ case class SQLEngineAppName(
       case _ => ZKPaths.makePath(s"${prefix}_$sharedLevel", user)
     }
   }
+
+  def getZkLockPath(prefix: String): String = {
+    sharedLevel match {
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/SQLEngineAppName.scala:38]: remove this `match-case`

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/HighAvailabilityConf.scala [hunk 6]
```diff
@@ -99,4 +99,11 @@ object HighAvailabilityConf {
     .timeConf
     .checkValue(_ > 0, "Must be positive")
     .createWithDefault(Duration.ofSeconds(120).toMillis)
+
+  val HA_ZK_ENGINE_LOCK_TIMEOUT: ConfigEntry[Long] =
+    buildConf("ha.zookeeper.engine.lock.timeout")
+      .doc("The timeout(s) of a lock when a engine created")
```
Comment (pan3793) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/HighAvailabilityConf.scala:105]: **when a engine created** => **when creating an engine**

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala (hunk 4 above)
Comment (pan3793) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:352]: yeap

Comment (Honglun) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:352]: `scala.util.Using` was introduced in scala-2.13 ?

Comment (pan3793) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:352]: we back ported it

Comment (Honglun) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:352]: Can we use this instead ?
  `def using[A <: {def close(): Unit}](r: => A)(f: A => Unit): Unit = {`
    `val resource = r`
    `try { f(resource) } finally { if (resource != null) r.close() }`
  `}`

Comment (Honglun) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:352]: `InterProcessSemaphoreMutex` is not a AutoCloseable, so  we need to add this in object `Releasable` ? 
`type AutoReleasableIsReleasable = {def release() : Unit}`
`implicit object AutoReleasableIsReleasable extends Releasable[{def release() : Unit}] {`
      `def release(resource: AutoReleasableIsReleasable): Unit = resource.release()`
`    }`
Or is there a better way? @pan3793

Comment (pan3793) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:352]: Oops, I didn't realize that `InterProcessSemaphoreMutex` is not a `AutoCloseable`, the extension is feasible, but don't do it(at least don't modify in `Using.scala` file), keep the `Using` compatible with future Scala versions.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala [hunk 7]
```diff
@@ -342,6 +343,31 @@ object ServiceDiscovery extends Logging {
     }
     serviceNode
   }
+
+  def withLock(
+      zkClient: CuratorFramework,
+      lockPath: String,
+      lockTimeOutMilliseconds: Long)(f: => Unit): Unit = {
```
Comment (pan3793) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:350]: Rename `lockTimeOutMilliseconds` => `lockTimeout`.
Personally, I agree with that any time variable should indicate the unit of values, but kyuubi already choose another style, all time should represent by `Long` in milliseconds, let's keep same style.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala (hunk 4 above)
Comment (pan3793) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:352]: Just leave it as it is.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala (hunk 7 above)
Comment (Honglun) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala]: Update

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/SQLEngineAppName.scala [hunk 8]
```diff
@@ -32,6 +32,13 @@ case class SQLEngineAppName(
       case _ => ZKPaths.makePath(s"${prefix}_$sharedLevel", user)
     }
   }
+
+  def getZkLockPath(prefix: String): String = {
+    assert(sharedLevel != CONNECTION)
+    sharedLevel match {
+      case _ => ZKPaths.makePath(s"${prefix}_$sharedLevel", "lock", user)
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/engine/SQLEngineAppName.scala:39]: `/kyuubi/kyuubi_USER/lock/userName/lock/lockes`, is the first  above `lock` neccessary?

Comment (Honglun) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/SQLEngineAppName.scala:39]: Yes. `GetServiceNodesInfo` cannot touch the `lock` node.
https://github.com/NetEase/kyuubi/blob/master/kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala#L266

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala [hunk 9]
```diff
@@ -342,6 +343,31 @@ object ServiceDiscovery extends Logging {
...
+        case e: Exception => throw KyuubiSQLException(s"Lock failed on path [$lockPath]", e)
+      }
+      f
+    } finally {
+      try {
+        if (lock != null) {
+          lock.release()
+        }
+      } catch {
+        case _: Exception =>
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:366]: What if we fail to release the lock？So it locks forever？

Comment (Honglun) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:366]: Yes, unless we restart `Kyuubi` Server. In this case, the new process acquires the lock after waiting for `HA_ZK_ENGINE_LOCK_TIMEOUT`.

Comment (yaooqinn) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:366]: I see. Seems we can use `ENGINE_INIT_TIMEOUT` for the lock timeout?

Comment (Honglun) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:366]: Yes, the default value of `HA_ZK_ENGINE_LOCK_TIMEOUT` is the same as `ENGINE_INIT_TIMEOUT`. Using 'ENGINE_INIT_TIMEOUT' should work. Do we need to modify this?

Comment (yaooqinn) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:366]: yes, let's use `ENGINE_INIT_TIMEOUT ` to make it simple

Comment (Honglun) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:366]: Update

Comment (yaooqinn) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:366]: thanks

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/SQLEngineAppName.scala (hunk 8 above)
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/engine/SQLEngineAppName.scala:39]: `ZKPaths.makePath(s"${prefix}_$sharedLevel", "lock", user) `？

File: kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala [hunk 10]
```diff
@@ -178,6 +178,7 @@ class KyuubiOperationManager private (name: String) extends OperationManager(nam
         if (remoteHandle != null) {
           val or = FetchOrientation.toTFetchOrientation(order)
           val req = new TFetchResultsReq(remoteHandle, or, maxRows)
+          req.setFetchType(1.toShort)
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala:181]: Is this related? can we rebase master branch?

**Review by yaooqinn — APPROVED** [review]
LGTM only minor comments

**Comment by yaooqinn** [comment]
Hi @Honglun  can you address the last few comments? I'd like to merge this

**Review by Honglun — COMMENTED** [review]
Hi @yaooqinn Any other questions about this?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala (hunk 10 above)
Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala:181]: I mean you need rebase master branch

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/SQLEngineAppName.scala (hunk 8 above)
Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/SQLEngineAppName.scala:39]: Hi @Honglun didn't you miss this comment? why do we need to match with _ here?

Comment (Honglun) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/SQLEngineAppName.scala:39]: I see, I will update this.

File: kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala (hunk 10 above)
Comment (Honglun) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala:181]: @yaooqinn help, can we have a chat in Gitter?

**Comment by Honglun** [comment]
Create a new PR
